it's sometimes very convenient to control SONAME

alexpokotilo opened the pull request:
after change made by https://github.com/Haivision/srt/pull/1741 our product cannot use both 1.4.2 and 1.4.3 even though there are no any ABI incompatibilities between them.
I understand why https://github.com/Haivision/srt/pull/1741 is made and agree it should be enabled by default but request some flexibility in SONAME assignment

maxsharabayko commented:
Maybe it is worth including `libsrt.so.1` just by default in addition to others. 🤔 
Thus making four `so`s in total.

- `libsrt.so.1.4.3`
- `libsrt.so.1.4`-> `libsrt.so.1.4.3`
- `libsrt.so.1`-> `libsrt.so.1.4.3`
- `libsrt.so` -> `libsrt.so.1.4.3`

alexpokotilo commented:
> Maybe it is worth including `libsrt.so.1` just by default in addition to others. thinking
> Thus making four `so`s in total.
> 
>     * `libsrt.so.1.4.3`
> 
>     * `libsrt.so.1.4`-> `libsrt.so.1.4.3`
> 
>     * `libsrt.so.1`-> `libsrt.so.1.4.3`
> 
>     * `libsrt.so` -> `libsrt.so.1.4.3`
Would be cool but I don't know how to add second "SONAME" to dynamic library. I actually think it's not possible.
The problem is not in creating new symlink. Of course, we can always add new symlink to libsrt.so.1.4.3 and name it libsrt.so.1. But usually dynamic libraries delivered without any links and then installer run "ldconfig" tool. This tool(at least on Linux) creates symlinks for every dynamic libraries in it's paths(configured in /etc/ld.so.conf).

Before #1741 fix, SRT library's SONAME was libsrt.so.1 and ldconfig created symlink libsrt.so.1 pointed into libsrt.so.1.4.2.
After #1741 fix, SRT library's SONAME is libsrt.so.1.4 and ldconfig creates symlink libsrt.so.1.4 pointed into libsrt.so.1.4.3.
There is no way to add second SONAME into dynamic library. Adding new symlink during build-time will not help here as library packages usually deliver only dynamic libraries without symlinks.

Usual/Common practice for Linux is to have SONAME with major version only(like is was in SRT before #1741). E.g:
 0x0000000000000001 (NEEDED)             Shared library: [libasan.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libssl.so.1.1]
 0x0000000000000001 (NEEDED)             Shared library: [libcrypto.so.1.1]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libxml2.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libzip.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]

As you can see only ssl libraries(libssl.so.1.1 and libcrypto.so.1.1) step away from this convention.
Our team use SRT library using dlopen. We load every function by dlname and if any function we need doesn't exists our product just unload library and log this as error. We want to work with 1.4.1, 1.4.2, 1.4.3 without complex logic of loading the library.
If SRT moderator team decided to use "ssl like" major.minor SONAME(which is no such a standard in Linux world) I understand that, but why don't we add ability to keep SONAME just MAJOR as it used to be ?

rndi commented:
Hi @alexpokotilo ,

Do you think you can explain a bit how this patch will help (you) with the unfortunate dynamic linkage breakage?

The way I see it is that if your program has to work with the libsrt library already installed on the users' machines, it has to compensate for 'libsrt.so.1.4' name change, my guess by trying dlopen() 'libsrt.so.1' first and then 'libsrt.so.1.4' as a fallback. And if you ship libsrt packaged with your program, perhaps you can adjust the name in your installer/builder?

I understand the desire for a quick solution for the current 1.4.x naming mess while ABI was not broken, but I am afraid this patch would need to be rolled back in 1.5.x where ABI is likely to break and it no longer makes sense to have an option/workaround to allow building with 'libsrt.so.1' SONAME.

alexpokotilo commented:
> Hi @alexpokotilo ,
> 
> Do you think you can explain a bit how this patch will help (you) with the unfortunate dynamic linkage breakage?

At least I will try to explain
> 
> The way I see it is that if your program has to work with the libsrt library already installed on the users' machines, it has to compensate for 'libsrt.so.1.4' name change, my guess by trying dlopen() 'libsrt.so.1' first and then 'libsrt.so.1.4' as a fallback. And if you ship libsrt packaged with your program, perhaps you can adjust the name in your installer/builder?

Yes, I can adjust library name using local CMakelist.txt fix. But I decided to add ability to use previous SONAME convention and postpone future incompatibility problems till 1.5. I think some others SRT users may decide to use libsrt.so.1 as SONAME till 1.5.
It's not that 1.5 is the next release. The next one will be 1.4.4 and libsrt.so.1 is fine for it too.

> 
> I understand the desire for a quick solution for the current 1.4.x naming mess while ABI was not broken, but I am afraid this patch would need to be rolled back in 1.5.x where ABI is likely to break and it no longer makes sense to have an option/workaround to allow building with 'libsrt.so.1' SONAME.

This pull request is all about "...in 1.5.x where ABI is likely to break".
We don't know for sure whether is will break ABI or not in the future but prevent using 1.4.2 and 1.4.3 in the same application while 1.4.2 and 1.4.2 are ABI compatible.

It would be great to postpone such changes till ABI incompatibility to prevent things like https://github.com/Haivision/srt/issues/1986 but 1.4.2 already released with such change.
What I propose, is to provide ability for package maintainers and end users of dozen of products to choose what 1.4.x version to use in their applications right now. Not in the future when SRT may change ABI. or may not. 


Here is SRT milestones https://github.com/Haivision/srt/milestones

The next release will be 1.4.4 not 1.5.0 and hence we have at least one year till 1.5.0 released, based on previous release frequency. Maybe a little less. Not too short life for for little CMake parameter, ha:).

But to be honest I'm not happy with this proposal too the same way as I am not happy with current SONAME approach released in 1.4.3.

What I actually want is above:)
In the ideal world SONAME must reflect API compatibility. From another hand @maxsharabayko declared ABI stability as a desired future goal https://github.com/Haivision/srt/issues/1598#issuecomment-805658608 and I agree with him.

What I want to emphasize by this pull request is that SONAME is not the same as SRT version. SONAME is for ABI compatibility.
SRT version from another hand is SRT milestone.
E.g we already know we will have 1.6.0 release. We don't know whether it will be ABI compatible with 1.5 or not, but current method will force application rebuilt to support 1.6. But what if some customers want to work with "old-and-good" 1.5 while others are ready to check out new 1.6. Even if 1.5 and 1.6 are fully ABI backward-compatible we will not be able to provide single binary program that can use both libraries.

The most interesting part is that so-called incompatibility is always(from my experience) related to CPerfMon structure.

I was able to find only one issue in 1.5 related to incompatibility and it's again related to CPerfMon
- https://github.com/Haivision/srt/issues/1138 

I think it's time to do the following:
- don't touch CPerfMon at all and keep it's current api functions srt_bstats and srt_bistats as they are.
- During #1138 implementation, Introduce new CPerfMon2 and work with it in SRT library internally.
- If application calls srt_bstats/srt_bistats, convert CPerfMon2 into CPerfMon and return CPerfMon for caller
- introduce new functions
 srt_bstats2(SRTSOCKET u, SRT_TRACEBSTATS2 * perf, int perf_size, int clear);
srt_bistats2(SRTSOCKET u, SRT_TRACEBSTATS2 * perf, int perf_size, int clear, int instantaneous);
- inside SRT library check perf_size and copy only "perf_size" bytes of SRT_TRACEBSTATS2 structure. if provided  perf_size > sizeof(SRT_TRACEBSTATS2), this means caller was compiled with later SRT version. In this case just fill perf_size - sizeof(SRT_TRACEBSTATS2) at the end of provided perf with zeros.

In this case if we add new fields into SRT_TRACEBSTATS2 nobody will be affected. If app built with new SRT is run with older library, caller will get 0 filled values for fields old library doesn't yet support.
It was initially bad design not to provide SRT_TRACEBSTATS's size. It's not too late to add it.

Back to SONAME.
I'd propose to set SONAME as constants in CMakeList.txt. E.g I propose to set SONAME in master as libsrt.so.1. We don't need to get it from SRT version. Once we introduce new API or change existing API, we will change it manually to libsrt.so.2 and so on.
There is no any relation between SRT version and it's SONAME. SRT version is the milestone, while SONAME is ABI compatibility version.
Of course, it would have been great if SRT milestone major version were changed depending on ABI.
But since SRT library version is a milestone for planning/marketing reasons, let is just keep SONAME the way it should be.
If we don't introduce any ABI changes between 1.5 and 1.7, why cannot we use use all range of srt libraries with single compiled application ?

maxsharabayko commented:
@alexpokotilo Thank you for sharing your thoughts on the subject! Indeed, one of the main ABI changer is `srt_bstats` function. As you correctly pointed out, the new versioning scheme was motivated by #1592 , e.g. the following comment by @FedericoCeratto:

> (...) ABI breaking changes without soname changes are impact Linux distributions e.g. [1] making srt difficult to ship. (...)
[1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=971754

@alexpokotilo you write:
> E.g we already know we will have 1.6.0 release. We don't know whether it will be ABI compatible with 1.5 or not, but current method will force application rebuilt to support 1.6. But what if some customers want to work with "old-and-good" 1.5 while others are ready to check out new 1.6. Even if 1.5 and 1.6 are fully ABI backward-compatible we will not be able to provide single binary program that can use both libraries.

I see your point here. A counter-argument might be that if the minor version has changed, there are likely some changes worth retesting. If API/ABI stays compliant, the only thing required is to rebuild the app against the newer SRT version.

The tricky part here is that e.g. v1.4.2 increased the `CBytePerfMon` structure. So if someone would consider seamlessly update SRT from v1.4.1 to v1.4.2 with the same `libsrt.so.1` dynamic library and without rebuilding the project, it can result in a crash or undefined behavior. Therefore, it might be safer to just don't let to find the `libsrt.so.1` library if wrong SRT version is built.
Those who want to ignore this can just reference `libsrt.so`.

Below is the list of ABI/API changes in the two latest releases. To proceed, we need to clearly state and agree on this versioning model and API/ABI compliance rules (to be done in #1602).

## API/ABI Compliance Rules

Assuming we proceed with the new `so` versioning, here are the rules.

1. **Extending API with new functions**. Can be done in patch versions. We assume that the user must check SRT version in runtime (using [`srt_version()`](https://github.com/Haivision/srt/blob/master/docs/API/API-functions.md#srt_getversion) function) and ensure the function exists in the dynamic library.
2. **Removing deprecated or old functions**. Can be done only in minor releases, e.g. `v1.4.3` -> `v1.5.0`.
3. **Changing existing API functions** interface. Not allowed. Can only remove, see no. 2.
4. **Removing items/fields from existing API structures** (e.g. `CBytePerfMon`). Not allowed. _Or allow in the minor release?_
5. **Adding items/fields to existing API structures** (e.g. `CBytePerfMon`). Allowed in minor releases (e.g. v1.4.* => v1.5.*). Allowed in patch releases only if it can stay compliant with the API within the minor version (e.g. size of the structure is passed as an argument to all functions and it can be used to determine the filling procedure).

To be done in #1602.

## Known ABI/API Changes

### API/ABI/Integration Changes in v1.4.3

- PR #1925: fixed/changed cast to `bool` instead of `int` in `srt_getsockopt(..)` / `srt_getsockflag(..)` API functions for:
   - [`SRTO_SENDER`](https://github.com/Haivision/srt/blob/v1.4.3/docs/API/API-socket-options.md#SRTO_SENDER),
   - [`SRTO_TSBPDMODE`](https://github.com/Haivision/srt/blob/v1.4.3/docs/API/API-socket-options.md#SRTO_TSBPDMODE),
   - [`SRTO_DRIFTTRACER`](https://github.com/Haivision/srt/blob/v1.4.3/docs/API/API-socket-options.md#SRTO_DRIFTTRACER),
   - [`SRTO_ENFORCEDENCRYPTION`](https://github.com/Haivision/srt/blob/v1.4.3/docs/API/API-socket-options.md#SRTO_ENFORCEDENCRYPTION).
- PR #1794: Fixed ABI compatibility around `SRTO_BINDTODEVICE` value depending on `ENABLE_EXPERIMENTAL_BONDING`.
- PR #1741: Made SRT versioned SO named with major and minor (now libsrt.so.1.4 instead of libsrt.so.1).

### API/ABI Changes in v1.4.2

- [`SRT_MSGCTRL.srctime`](https://github.com/Haivision/srt/blob/v1.4.2/srtcore/srt.h#L862) now expects SRT internal clock timestamp (as returned from srt_time_now()). In previous versions a wallclock time in microseconds since the UNIX epoch was expected.
- Several fields were added to the [`CBytePerfMon` structure](https://github.com/Haivision/srt/blob/v1.4.2/srtcore/srt.h#L402) (statistics).

rndi commented:
> I propose to set SONAME in master as libsrt.so.1. We don't need to get it from SRT version. Once we introduce new API or change existing API, we will change it manually to libsrt.so.2 and so on.

Yes, fully agree, that's the way it should be. But I don't know if it is possible to put libsrt.1.4 SONAME jiny back in the bottle now that's it is out: what may happen next is that external projects like ffmpeg, gstreamer, etc will start adjusting to the new naming only for us to break them again with roll back to libsrt.1 SONAME. On top of that, if your patch is applied to master and the other build external systems start to make use of it, this will break again in unpleasant ways when we do change the ABI and either remove or forget to remove this patch.

The ideal workaround, as it has been already mentioned, would be to somehow resume distributing "libsrt.1.so" along with "libsrt.1.4.so" or recalling "libsrt.1.4.so". Supplying two libraries with different SONAMEs for the period of 1.4.x releases maybe? - probably a horrible idea too. Declaring 1.4.4 a flop and releasing 1.4.5 immediately? - don't know how well that may fly. Failing that I would vote for sticking with annoying consistency (libsrt.1.4) than switching things back and forward.

maxsharabayko commented:
Please remember that strictly speaking the API/ABI was already partially broken (see #1592 and [my above comment](https://github.com/Haivision/srt/pull/1983#issuecomment-833402441)).
`libsrt.so.1` is used e.g. in Ubuntu as SRT v1.4.0 API/ABI (see [here](https://packages.ubuntu.com/focal/libsrt1)).
Following this logic, there is already no `libsrt.so.1`, but `libsrt.so.2`.

`libsrt.so.1.4` seems to be already in use [here](https://packages.ubuntu.com/hirsute/libsrt1.4-openssl) to address this issue.

alexpokotilo commented:
> `libsrt.so.1.4` seems to be already in use [here](https://packages.ubuntu.com/hirsute/libsrt1.4-openssl) to address this issue.

I checked @rndi's points and this reply and agree that we already made some mistakes and shouldn't do new.
- let's reject this pull request and I'll propose new one there 'libsrt.so.1.4' will current SONAME.
- let's change SONAME every time manually per "API/ABI Compliance Rules" described https://github.com/Haivision/srt/pull/1983#issuecomment-833402441
- once ABI changed in incompatible way SONAME is changed manually to the next value regardless of the next milestone. 
E.g if we add new api between 1.4.3 and 1.4.4 SONAME will be changed into 1.5.
- if we keep current method of SONAME definition we will not be able to introduce new API or change any structure between minor releases.
So I vote to unbind SONAME and SRT milestone or we will not be able to follow "API/ABI Compliance Rules" described https://github.com/Haivision/srt/pull/1983#issuecomment-833402441

alexpokotilo commented:
> @alexpokotilo you write:
> 
> > E.g we already know we will have 1.6.0 release. We don't know whether it will be ABI compatible with 1.5 or not, but current method will force application rebuilt to support 1.6. But what if some customers want to work with "old-and-good" 1.5 while others are ready to check out new 1.6. Even if 1.5 and 1.6 are fully ABI backward-compatible we will not be able to provide single binary program that can use both libraries.
> 
> I see your point here. A counter-argument might be that if the minor version has changed, there are likely some changes worth retesting. 

It's always good idea to test application with new library, that is true. But new SONAME should signal that changes was made but not that it was probably made.
> If API/ABI stays compliant, the only thing required is to rebuild the app against the newer SRT version.

not true unfortunately. If you build application with newer SRT version e.g 1.6, your application will require libsrt.so.1.6 and will not work with 1.5 libraries since they have SONAME libsrt.so.1.5. Linker requires exact library libsrt.so.1.6 and will not load libsrt.so.1.5. So you will have to provide two binary version of your application to be able to work with 1.6 and 1.5.

Let me сite what @sdroege said here https://github.com/Haivision/srt/issues/1592#issuecomment-757991864
"You have to increase that 1 there whenever the ABI is changing in incompatible ways. If you're lazy you could simply increase it with every release."(c)
It's brilliant clear and precise. We are not lazy and I propose to change SONAME manually the way I described above.
The main problem with current approach is that SRT library is not as stable as a stone. It's not quite rate to have serious bugs in new releases. Library is still under very active development so it's expected to have such cases. But I want to provide our customers with ability to choose what compatible srt version to install. If they want the new one, they use it. If they want to rollback to previous release, that is fine too.
I'm arguing against auto-assigning SONAME based on milestone. Do we really sure all future SRT minor releases will be ABI incompatible? Do we really ready to create new minor release once we need to change an ABI urgently ? 
We can start SONAME from 1.4 now for historical reason and increase it by one between every patch releases or on the next minor release only if necessary. I mean if we need to update ABI between 1.5.1 and 1.5.2 several times, we increase current SONAME just once.  
From another hand it's highly unlikely to have backward and forward ABI compatibility between minor releases. That's true.

> 
> The tricky part here is that e.g. v1.4.2 increased the `CBytePerfMon` structure. So if someone would consider seamlessly update SRT from v1.4.1 to v1.4.2 with the same `libsrt.so.1` dynamic library and without rebuilding the project, it can result in a crash or undefined behavior. Therefore, it might be safer to just don't let to find the `libsrt.so.1` library if wrong SRT version is built.

here is what I did I my core to work with both 1.4.1 and 1.4.2
```

// every new SRT versions adds new fields into SRT_TRACEBSTATS structure hence "srt_bstats" and "srt_bstats"
// functions overwrite SRT_TRACEBSTATS structure Nimble compiled with. Adding "safe_guard_boundary" after "stats"
// prevents panic in this case
struct SRT_TRACEBSTATS_WRAPPER {
    SRT_TRACEBSTATS stats;
    uint8_t safe_guard_boundary[1024];
};

int srt_bstats_safe(SRTSOCKET u, SRT_TRACEBSTATS* perf, int clear) {
    SRT_TRACEBSTATS_WRAPPER stats_wrapper;
    int ret = LibSrt::srt_bstats(u, &stats_wrapper.stats, clear);
    memcpy(perf, &stats_wrapper.stats, sizeof(SRT_TRACEBSTATS));
    return ret;
}
int srt_bistats_safe(SRTSOCKET u, SRT_TRACEBSTATS * perf, int clear, int instantaneous) {
    SRT_TRACEBSTATS_WRAPPER stats_wrapper;
    int ret = LibSrt::srt_bistats(u, &stats_wrapper.stats, clear, instantaneous);
    memcpy(perf, &stats_wrapper.stats, sizeof(SRT_TRACEBSTATS));
    return ret;
}

```

this is workaround to silly api. It's very simple to add srt_bstats2 and srt_bstats2  as I previously proposed and set only provided part of structure.
1) if client's provided structure is longer than current's SRT stats structure(application compiled with newer SRT version), then zeros delta bytes of client structure and copy SRT internal stats into client's provided structure completely
2) if client's provided structure is less than SRT internal stats structure(application compiled with older srt version) then, copy only as much bytes as client's structure is.

Look, all we have to do is to add **size_t perf_size** parameter along with SRT_TRACEBSTATS2* perf. That is it. It's so easy and keep us away from these stats dances altogether. As you can see in my code above I've already did the same on application side and this works on all platforms including Android/IOS.

> 
> Below is the list of ABI/API changes in the two latest releases. To proceed, we need to clearly state and agree on this versioning model and API/ABI compliance rules (to be done in #1602).

Holy truth
> ## API/ABI Compliance Rules
> 
> Assuming we proceed with the new `so` versioning, here are the rules.

we have to proceed with it now as it have been introduced.
> 
>     1. **Extending API with new functions**. Can be done in patch versions. We assume that the user must check SRT version in runtime (using [`srt_version()`](https://github.com/Haivision/srt/blob/master/docs/API/API-functions.md#srt_getversion) function) and ensure the function exists in the dynamic library.

Great idea! Keep ABI backward compatible but add ability to add new functions. And we don't have to change SONAME this way. It's against common Linux practices though as any ABI changes should be reflected in SONAME. I think open-source libraries will not be happy with this method.
E.g ffmpeg library. It's current version is e.g 4.3.1. But it's libraries SONAMEs are:

- libavutil.so.56
- libavcodec.so.58
- libavfilter.so.7
- libswresample.so.3
- libswscale.so.5

there is no any relation between ffmpeg milestone version 4.3.1 and it's libraries versions. Moreover there are no relations between libraries version in different libraries. Once ABI changed, library version changes as well.
This is a real-life example of the case glass to glass similar to SRT.
ffmpeg has milestone version and ABI version. SRT just started it's ABI version from 1.4. Well, yeh strange, but it's how it is.

> 
>     2. **Removing deprecated or old functions**. Can be done only in minor releases, e.g. `v1.4.3` -> `v1.5.0`.
> 

Agree.

>     3. **Changing existing API functions** interface. Not allowed. Can only remove, see no. 2.
> 

Totally agree!

>     4. **Removing items/fields from existing API structures** (e.g. `CBytePerfMon`). Not allowed. _Or allow in the minor release?_

Never allowed. 
> 
>     5. **Adding items/fields to existing API structures** (e.g. `CBytePerfMon`). Allowed in minor releases (e.g. v1.4.* => v1.5.*). Allowed in patch releases only if it can stay compliant with the API within the minor version (e.g. size of the structure is passed as an argument to all functions and it can be used to determine the filling procedure).

If we don't remove/change types of statistics field and don't remove old fields then we can add new fields any time.

I see you checked my proposals about stats size. Thank you ! I really think we need to stop these dances with stats. SRT moderators made so great and complex work so it is really worth stop a little and re-think this stats. it's so easy to fix this and forget about it. Please review deeply my comments about stats api.  We need to introduce new CBytePerfMon2 and replace **int**s with wider type like uint64_t in advance. then we should add new fields at the end of CBytePerfMon2. and don't touch useless fields at all.
Old stats functions will just convert CBytePerfMon2 into CBytePerfMon for these fields presented in CBytePerfMon.
Providing CBytePerfMon2 size will do all the rest dirty job. Callers with newer and hence bigger CBytePerfMon2 will get new fields if they exist while callers with shorter CBytePerfMon2 will get what they expected.
Old-fashioned guy calling previous stats with CBytePerfMon without size, will no longer have any problems if they use 1.4.2 or newer.
it is a paradise, isn't it ?

> 
> 
> To be done in #1602.
> ## Known ABI/API Changes
> ### API/ABI/Integration Changes in v1.4.3
> 
>     * PR #1925: fixed/changed cast to `bool` instead of `int` in `srt_getsockopt(..)` / `srt_getsockflag(..)` API functions for:
>       
>       * [`SRTO_SENDER`](https://github.com/Haivision/srt/blob/v1.4.3/docs/API/API-socket-options.md#SRTO_SENDER),
>       * [`SRTO_TSBPDMODE`](https://github.com/Haivision/srt/blob/v1.4.3/docs/API/API-socket-options.md#SRTO_TSBPDMODE),
>       * [`SRTO_DRIFTTRACER`](https://github.com/Haivision/srt/blob/v1.4.3/docs/API/API-socket-options.md#SRTO_DRIFTTRACER),
>       * [`SRTO_ENFORCEDENCRYPTION`](https://github.com/Haivision/srt/blob/v1.4.3/docs/API/API-socket-options.md#SRTO_ENFORCEDENCRYPTION).
> 
>     * PR #1794: Fixed ABI compatibility around `SRTO_BINDTODEVICE` value depending on `ENABLE_EXPERIMENTAL_BONDING`.

Well, as you stated in #1925 nobody actually affected as it's hard to imagine why to ask library to provide SRTO_SENDER or SRTO_TSBPDMODE if you previously set them. But formally it's ABI change and we have changed ABI version into 1.4 :)

> 
>     * PR #1741: Made SRT versioned SO named with major and minor (now libsrt.so.1.4 instead of libsrt.so.1).
 
We had do that. Version should we libsrt.so.2 but it's how it is. We can change it to libsrt.so.2 in the next ABI change.

> 
> ### API/ABI Changes in v1.4.2
> 
>     * [`SRT_MSGCTRL.srctime`](https://github.com/Haivision/srt/blob/v1.4.2/srtcore/srt.h#L862) now expects SRT internal clock timestamp (as returned from srt_time_now()). In previous versions a wallclock time in microseconds since the UNIX epoch was expected.

only people using SRT_MSGCTRL affected. But before 1.4.2 SRT_MSGCTRL's clock time doesn't work. So nobody affected actually

> 
>     * Several fields were added to the [`CBytePerfMon` structure](https://github.com/Haivision/srt/blob/v1.4.2/srtcore/srt.h#L402) (statistics).

Well, I've already said too much about stats, you know:)

gizahNL commented:
The normal way to do it is to decouple soversion (=ABI version) from library/API version.
As described here:
https://github.com/pvanhoof/dir-examples

maxsharabayko commented on CMakeLists.txt:
```diff
@@ -151,6 +151,7 @@ option(USE_OPENSSL_PC "Use pkg-config to find OpenSSL libraries" ON)
 option(USE_BUSY_WAITING "Enable more accurate sending times at a cost of potentially higher CPU load" OFF)
 option(USE_GNUSTL "Get c++ library/headers from the gnustl.pc" OFF)
 option(ENABLE_SOCK_CLOEXEC "Enable setting SOCK_CLOEXEC on a socket" ON)
+option(ENABLE_SRT_VERSIONED_SO_WITH_MAJOR_MINOR "Make SRT versioned SO named with SONAME containing major and minor versions" ON)
```
```suggestion
option(ENABLE_SRT_SONAME_SUFFIX_MAJOR_ONLY "Make SRT SONAME suffixed with major version only (not recommended)" OFF)
```

maxsharabayko commented on CMakeLists.txt:
```diff
@@ -893,7 +894,11 @@ if (srt_libspec_shared)
 	set (CMAKE_POSITION_INDEPENDENT_CODE ON)
 	set_property(TARGET ${TARGET_srt}_shared PROPERTY OUTPUT_NAME ${TARGET_srt})
-	set_target_properties (${TARGET_srt}_shared PROPERTIES VERSION ${SRT_VERSION} SOVERSION ${SRT_VERSION_MAJOR}.${SRT_VERSION_MINOR})
+	if (ENABLE_SRT_VERSIONED_SO_WITH_MAJOR_MINOR)
+		set_target_properties (${TARGET_srt}_shared PROPERTIES VERSION ${SRT_VERSION} SOVERSION ${SRT_VERSION_MAJOR}.${SRT_VERSION_MINOR})
+	else ()
+		set_target_properties (${TARGET_srt}_shared PROPERTIES VERSION ${SRT_VERSION} SOVERSION ${SRT_VERSION_MAJOR})
+	endif()
```
```suggestion
	if (ENABLE_SRT_SONAME_SUFFIX_MAJOR_ONLY)
		set_target_properties (${TARGET_srt}_shared PROPERTIES VERSION ${SRT_VERSION} SOVERSION ${SRT_VERSION_MAJOR})
	else ()
		set_target_properties (${TARGET_srt}_shared PROPERTIES VERSION ${SRT_VERSION} SOVERSION ${SRT_VERSION_MAJOR}.${SRT_VERSION_MINOR})
	endif()
```

maxsharabayko reviewed (changes requested):
The [Build Options doc](https://github.com/Haivision/srt/blob/master/docs/build/build-options.md) has to be updated as well:

```markdown
#### ENABLE_SRT_SONAME_SUFFIX_MAJOR_ONLY
**`--enable-srt-soname-suffix-major-only`** (default: OFF)

The patch version upgrade usually means improvements and/or bug fixes, that do not (notably) affect SRT API/ABI compatibility and do not significantly change the behavior. Thus, an application can relatively safely update onto the next SRT patch release.

Any minor version upgrade means SRT might no longer be fully compliant with the OLD API and/or old behavior or contains notable changes in the source code. In this case, an application integrating SRT should consider upgrading to the new minor release as a major step, requiring proper testing and API usage review.

By default the SRT SO/DLL name is suffixed with `major.minor` version:
- `libsrt.so.1.4`-> `libsrt.so.1.4.3`.
- `libsrt.so.1.4`-> `libsrt.so.1.4.4`
- `libsrt.so.1.5`-> `libsrt.so.1.5.0`
- ...

Enabling this build option changes the SRT SO/DLL name suffix to `major` version only:
- `libsrt.so.1.4`-> `libsrt.so.1`.
- `libsrt.so.1.4`-> `libsrt.so.1`
- `libsrt.so.1.5`-> `libsrt.so.1`
- ...

This should only be done with full awareness of the intent, and should only help to transition existing solutions onto the new naming convention.
```

maxsharabayko commented:
It's been a long since the PR had been proposed. And there seems to be not too much feedback or voting for it.
@alexpokotilo is it still topical to you?

alexpokotilo commented:
@maxsharabayko 
I think it is useful for people who wants to deliver products with multiple SRT version support.
A year has been passed and our product works with 1.4.1-1.4.4 versions and user can choose which one to use.
It is topical

alexpokotilo commented:
@maxsharabayko 
our product loads srt library dynamically at runtime(using dlopen) and can get current library version in run-time and process different versions behavior in application. This option is useful for such cases.
But most users use static or dynamic(without dlopen) linking and this is why there is no voting for this change.
Adding this flag set OFF by default with description you made should not hurt anyone.

maxsharabayko commented:
@alexpokotilo 
I see.
Just to better understand... By default the following is generated:
- llibsrt.so -> libsrt.so.1.4
- libsrt.so.1.4 -> libsrt.so.1.4.4
- libsrt.so.1.4.4

What is the benefit of having `libsrt.so.1` over just using the `libsrt.so` if you don't care about the API version?

alexpokotilo commented:
llibsrt.so -> libsrt.so.1.4 is generated only during building. Then you deliver libsrt.so.1.4.4 ldconfig generates only SYMLINK depending on your SONAME. if your SONAME is libsrt.so.1.4 you will get
libsrt.so.1.4 -> libsrt.so.1.4.4
libsrt.so.1.4.4
and no llibsrt.so will be generated
